stat/test/wilcoxon_hyp.go

```diff
@@ -0,0 +1,182 @@
+	for i := 0; i < lengthOfIn; i++ {
+		pos = pairStructArr[i].Position
+		val = pairStructArr[i].Value
+		if val != 0 {
+			out[pos] = float64(mpOut[val]) / float64(mpCount[val])
+		}
+		if mpCount[val] > 1 {
+			if _, ok := tmpRank[val]; !ok {
+				tmpRank[val] = struct{}{}
+			}
+		}
+	}
+	for val := range tmpRank {
+		tieAdjust += (mpCount[val] * mpCount[val] * mpCount[val]) - (mpCount[val])
+
+	}
+	return tieAdjust
+}
+
+func wilCoxonSignedRankTest(x, y []float64) (float64, int, int) {
+	z := make([]float64, len(x))
+	absZ := make([]float64, len(x))
+	floats.SubTo(z, x, y)
+
+	allZero := true
+	for _, val := range z {
+		if val != 0 {
+			allZero = false
+			break
+		}
+	}
+	if allZero {
+		panic(badAllElementsAreSame)
+	}
+
+	calculateAbsDifference(absZ, x, y)
+	ranks := make([]float64, len(x))
+	tieAdj := rank(ranks, absZ)
+	n := len(x)
+
+	WPlus := 0.0
+	WMinus := 0.0
+	for index, rank := range ranks {
+		if z[index] > 0 {
+			WPlus += rank
+		} else if z[index] == 0 {
+			n--
+		}
+	}
+	WMinus = (float64(n*(n+1)) / 2.0) - WPlus
+	return math.Max(WMinus, WPlus), n, tieAdj
+}
+
+func calculateExactPValue(Wmax float64, NZ int) float64 {
+	m := 1 << NZ
+	largerRankSums := 0
+
+	for i := 0; i < m; i++ {
+		rankSum := 0
+		// Generate all possible rank sums.
+		for j := 0; j < NZ; j++ {
+			// (i >> j) & 1 extract i's j-th bit from the right
+			if ((i >> j) & 1) == 1 {
+				rankSum += j + 1
+			}
+		}
+		if float64(rankSum) >= Wmax {
+			largerRankSums++
+		}
+	}
+
+	// largerRankSums / m gives the one-sided p-value, so it's multiplied
+	// with 2 to get the two-sided p-value.
+	return 2 * (float64(largerRankSums) / float64(m))
+}
+
+func calculateAsymptoticPValue(Wmin float64, NZ, tieAdj int) float64 {
+	// n should be number of non-zero absolute difference pairs
+	ES := float64(NZ*(NZ+1)) / 4.0
+	VarS := (ES * (float64(2*NZ + 1)) / 6.0) - (float64(tieAdj) / 48.0)
+
+	// - 0.5 is a continuity correction
+	z := (Wmin - ES - 0.5) / math.Sqrt(VarS)
+
+	standardNormal := distuv.UnitNormal
+	return 2 * standardNormal.CDF(z)
+}
+
+// WilcoxonSignedRankTest implements Wilcoxon signed rank test (https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test)
+// The Wilcoxon signed-rank test tests the null hypothesis that two paired samples come from the same distribution.
+// In particular, it tests whether the distribution of the differences x - y is symmetric about zero. It is a non-parametric version of the paired T-test.
+// x and y are the first and second set of measurements respectively. Exact P value computation is expensive, it is only available for the sample sizes with size less than 30.
+// WilcoxonSignedRankTest panics when exactPValue is true and len(x) is greater than 30. The two sided p-value for the test.
+func WilcoxonSignedRankTest(x, y []float64, exactPValue bool) float64 {
+	ensureDataConformance(x, y)
+	Wmax, N, tieAdj := wilCoxonSignedRankTest(x, y)
+	if exactPValue && N > 30 {
+		panic(BadNumberTooLarge)
+	}
+	if exactPValue {
```

Comment:
Would you consider also returning the test statistic? It is something the users will often also want to report when analysing data.